File: app/src/main/java/kim/uno/ui/holder/BaseViewHolder.java
package kim.uno.ui.holder;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import kim.uno.ui.adapter.BaseRecyclerAdapter;

public abstract class BaseViewHolder<T> extends RecyclerView.ViewHolder {

    public BaseRecyclerAdapter mAdapter;
    public Context mContext;

    public BaseViewHolder(BaseRecyclerAdapter adapter, ViewGroup parent, int resId) {
        this(adapter, LayoutInflater.from(adapter.getContext()).inflate(resId, parent, false));
    }

    public BaseViewHolder(BaseRecyclerAdapter adapter, View itemView) {
        super(itemView);
        mAdapter = adapter;
        mContext = mAdapter.getContext();
    }

    public void onBindView(final T item, final int position) { }

}

File: Base/src/main/kotlin/kim/uno/kotlin/base/ui/BaseViewHolder.kt
package kim.uno.kotlin.base.ui

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kim.uno.kotlin.base.item.RecyclerItem

abstract class BaseViewHolder<T : RecyclerItem>(open val adapter: BaseRecyclerAdapter<*>, itemView: View) : RecyclerView.ViewHolder(itemView) {

    constructor(adapter: BaseRecyclerAdapter<*>, parent: ViewGroup, resId: Int) : this(adapter, LayoutInflater.from(adapter.recyclerView!!.context).inflate(resId, parent, false))

    open fun onBindView(item: T?, position: Int) { }

    protected val context: Context = adapter.recyclerView!!.context
    open fun getItem(): T? = adapter.getItem(adapterPosition) as T?
    open fun onScrollChanged(position: Float, dx: Int, dy: Int) { }

}